docs/proposals/scheduling/20220609-resource-reservation.md

```diff
@@ -0,0 +1,234 @@
+	Phase ReservationPhase `json:"phase,omitempty"`
+	// The `Conditions` indicate the messages of reason why the reservation is still pending.
+	Conditions []ReservationCondition `json:"conditions,omitempty"`
+	// Current resource owners which allocated the reservation resources.
+	Owners []ReservationOwner `json:"owners,omitempty"`
+}
+
+type ReservationOwner struct {
+	// refer to a controller when the field `Controller` is set.
+	metav1.OwnerReference
+	// <unresolved>
+	// The `Namespace` field is needed when Reservation is not namespaced.
+	// </unresolved>
+	Namespace string `json:"namespace,omitempty"`
+}
+
+type ReservationOwnerSelector struct {
+	// Multiple field selectors are ORed.
+	ReservationOwner
+	LabelSelector *metav1.LabelSelector `json:"labelSelector,omitempty"`
+}
+
+type ReservationPhase string
+
+const (
+	ReservationPhasePending   ReservationPhase = "Pending"
+	ReservationPhaseAvailable ReservationPhase = "Available"
+	ReservationPhaseWaiting   ReservationPhase = "Waiting"
+	ReservationPhaseExpired   ReservationPhase = "Expired"
+)
+
+type ReservationCondition struct {
+	LastProbeTime      metav1.Time `json:"lastProbeTime"`
+	LastTransitionTime metav1.Time `json:"lastTransitionTime"`
+	Reason             string      `json:"reason"`
+	Message            string      `json:"message"`
+}
+```
+
+### Implementation Details
+
+#### Reservation Plugin
+
+A `Reservation` object has its resource requirements like a pod. Ideally, A `Reservation` object should get processed directly by the scheduler like a pod. However, it can require a series of modifications on [scheduling framework](#glossary), losing the compatibility with standard kube-scheduler, kubelet, autoscaler, etc. In the reservation plugin, we fake one *reservation pod* for one `Reservation` inside the scheduler to fulfill general scheduling plugins (noderesources, nodeaffinity, tainttolerations, ...). The scheduling framework can handle `Reservation` objects by processing fake pods in both [scheduling cycle and binding cycle](https://kubernetes.io/docs/concepts/scheduling-eviction/scheduling-framework/#scheduling-cycle-binding-cycle).
+
+A fake pod inside the scheduler can construct the same affinity/anti-affinity constraints as owner pods, which may change the reservation result. To handle this problem, koord-scheduler should extend the framework to skip check of pod affinity for existing reservations in the Filter phase.
+
+Once the scheduling decision has been made, the corresponding `Reservation` object is updated with a new status indicating whether the reservation succeeded or not. The fake pod does not expose to other components, and the kubelet without modification does not perceive a `Reservation` assigned. Fortunately, a `Reservation` does not need to be executable on the node, so existing containers can keep running as usual.
+
```

Comment:
There are still some questions about the life cycle of Reservation. Who is responsible for destroying the Reservation? If it expires, is the scheduler responsible for GC? If it does not expire, is it up to the user to decide how to delete it? In this scenario, what should we recommend users to do? It is recommended to append some relevant descriptions to clarify this responsibility and the GC mechanism.

Reply:
Add a section for GC.